master/internal/api_user.go

```diff
@@ -166,32 +167,37 @@ func (a *apiServer) PatchUser(
 	if err != nil {
 		return nil, err
 	}
-	if !curUser.Admin && curUser.ID != model.UserID(req.UserId) {
+	uid := model.UserID(req.UserId)
+	if !curUser.Admin && curUser.ID != uid {
 		return nil, grpcutil.ErrPermissionDenied
 	}
-	user := &model.User{ID: model.UserID(req.UserId)}
 	// TODO: handle any field name:
-	if req.User.DisplayName != "" {
-		user.DisplayName = null.StringFrom(req.User.DisplayName)
-		switch err = a.m.db.UpdateUser(user, []string{"display_name"}, nil); {
-		case err == db.ErrNotFound:
-			return nil, errUserNotFound
-		case err != nil:
-			return nil, err
+	if req.User.DisplayName != nil {
+		u := &userv1.User{}
+		if req.User.DisplayName.Value == "" {
+			err = a.m.db.QueryProto("set_user_display_name", u, req.UserId, "")
+		} else {
+			// Remove non-ASCII chars to avoid hidden whitespace, confusable letters, etc.
+			re := regexp.MustCompile("[^\\p{Latin}\\p{N}\\s]")
+			displayName := re.ReplaceAllLiteralString(req.User.DisplayName.Value, "")
+			// Restrict 'admin' and 'determined' in display names.
+			if !(curUser.Admin && curUser.ID == uid) && strings.Contains(strings.ToLower(displayName),
+				"admin") {
+				return nil, status.Error(codes.InvalidArgument, "Non-admin user cannot be renamed 'admin'")
+			}
+			if curUser.Username != "determined" && strings.Contains(strings.ToLower(displayName),
```

Comment:
question: does this need to use `strings.Contains`? or would having "admin" or "determined" as a substring be acceptable?